text/0000-grid-identity.md

```diff
@@ -0,0 +1,718 @@
+`AlternateIDIndexEntry` for that alternate ID is deleted) and any new IDs will
+be checked and added.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## State
+
+Agents, organizations, and roles are stored in state using protobuf message
+format. These messages are defined as follows:
+
+```protobuf
+message Agent {
+  string org_id = 1;
+  string public_key = 2;
+  bool active = 3;
+  repeated string roles = 4;
+  repeated KeyValueEntry metadata = 5;
+}
+
+message AgentList {
+    repeated Agent agents = 1;
+}
+
+message Organization {
+  string org_id = 1;
+  string name = 2;
+  repeated string locations = 3;
+  repeated AlternateID alternate_ids = 4;
+  repeated KeyValueEntry metadata = 5;
+}
+
+message OrganizationList {
+  repeated Organization organizations = 1;
+}
+
+message Role {
+  string org_id = 1;
+  string name = 2;
+  string description = 3;
+  repeated string permissions = 4;
+  repeated string allowed_organizations = 5;
+  repeated string inherit_from = 6;
+}
+
+message RoleList {
+  repeated Role roles = 1;
+}
+
+message AlternateIDIndexEntry {
+  string id_type = 1;
+  string id = 2;
+  string grid_identity_id = 3;
+}
+
+// Helpers
+message KeyValueEntry {
+  string key = 1;
+  string value = 2;
+}
+
+message AlternateID {
+  string id_type = 1;
+  string id = 2;
+}
+```
+
+## Transactions
+
+There is one `IdentityPayload` message which accounts for all Identity
+transactions. This message has an enum to indicate if the transaction is a
+creation, update, or delete action for each of the three core data types in
+Identity.
+
+```protobuf
+message IdentityPayload {
+  enum Action {
+    ACTION_UNSET = 0;
+    CREATE_AGENT = 1;
+    UPDATE_AGENT = 2;
+    DELETE_AGENT = 3;
+    CREATE_ORGANIZATION = 4;
+    UPDATE_ORGANIZATION = 5;
+    DELETE_ORGANIZATION = 6;
+    CREATE_ROLE = 7;
+    UPDATE_ROLE = 8;
+    DELETE_ROLE = 9;
+  }
+
+  Action action = 1;
+  CreateAgentAction create_agent = 2;
+  UpdateAgentAction update_agent = 3;
+  DeleteAgentAction delete_agent = 4;
+  CreateOrganizationAction create_organization = 5;
+  UpdateOrganizationAction update_organization = 6;
+  DeleteOrganizationAction delete_organization = 7;
+  CreateRoleAction create_role = 8;
+  UpdateRoleAction update_role = 9;
+  DeleteRoleAction delete_role = 10;
+}
+
+message CreateAgentAction {
+  string org_id = 1;
+  string public_key = 2;
+  bool active = 3;
+  repeated string roles = 4;
+  repeated KeyValueEntry metadata = 5;
+}
+
+message UpdateAgentAction {
+  string org_id = 1;
+  string public_key = 2;
+  bool active = 3;
+  repeated string roles = 4;
+  repeated KeyValueEntry metadata = 5;
+}
+
+message DeleteAgentAction {
+  string org_id = 1;
+  string public_key = 2;
+}
+
+message CreateOrganizationAction {
+  string id = 1;
+  string name = 2;
+  // List of IDs for Grid Locations associated with the organization
+  repeated string locations = 3;
```

Comment:
Locations should probably be excluded from CreateOrganizationAction. The locations will likely not exist at this point, since locations will be owned by the organization we are creating. We probably want a sequence like:

- create organization (without locations)
- create locations
- update organization w/locations